### PR TITLE
🕸️ Network library

### LINKED ISSUE #2: 🕸️ Network library
# 👮‍♂️ What's needed
Creating a networking library would be a great idea. This library would handle request sending and response awaiting on one part and on another handle routes, request receive and response sending.

## 📋 Subtasks
- [x] Create types
- [x] Create the router
- [x] Create a "Promise" handler for request sending

# 📜 Sources
We made a little hand drawing of what it could look like.
Quick sheets - page 5

### PR DESCRIPTION
**Pull request by axel-eck**
We made a network library that enables anyone to send request and make a router with no network understanding, we basically removed the two first software network layers.

Client side:
- Send request with a route, a method, optional parameters, optional JSON body
- Asynchronous "promise" type request sending, you provide what we call a consumer that will be automatically called when we received the response

Router (server) side:
- Create routes that can handle any route with a given method
- Create middlewares that can run and modify any route under them

Router example:
```C
#include <stdlib.h>
#include <string.h>
#include "network/router.h"

network_router_t *router_global = NULL;

response_t *router_handle_test(request_t *request)
{
    response_t *response = calloc(1, sizeof(response_t));
    char *body = "{\"message\": \"Hello, World!\"}";

    response->header.status_code = 200;
    response->header.content_length = strlen(body);
    response->body = body;
    return response;
}

response_t *router_handle_stop(request_t *request)
{
    response_t *response = calloc(1, sizeof(response_t));
    char *body = "{\"message\": \"Goodbye, World!\"}";

    response->header.status_code = 200;
    response->header.content_length = strlen(body);
    response->body = body;
    return response;
}

int main(void)
{
    host_t *host = malloc(sizeof(host_t));
    network_router_t *router = NULL;

    if (!host)
        return 84;
    host->port = 8080;
    host->ip = "127.0.0.1";
    router = network_router_create(*host);
    if (!router) {
        free(host);
        return 84;
    }
    router_global = router;
    network_router_add_route(router, (route_t) {GET, "/api/v1/test"}, &router_handle_test);
    network_router_add_route(router, (route_t) {POST, "/api/v1/stop"}, &router_handle_stop);
    network_router_listen(router);
    network_router_destroy(router);
    free(host);
    return 0;
}
```

Client example:
```C
#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include "network/networking.h"
#include "network/network_manager.h"

api_handler_t *handler_global = NULL;

static void handle_response(response_t *response)
{
    printf("Response received: %d %s\n", response->header.status_code, response->body);
    if (strstr(response->body, "Hello, World!") != NULL)
        network_send_request(handler_global, network_create_request_no_params_no_body((route_t) {POST, "/api/v1/stop"}), handle_response);
    else
        handler_global->running = false;
}

int main(int argc, char **argv)
{
    host_t *host = malloc(sizeof(host_t));
    api_handler_t *handler;
    request_t *request;

    if (argc != 3) {
        free(host);
        return 84;
    }
    host->port = (int) strtol(argv[1], NULL, 10);
    host->ip = argv[2];
    handler = network_create_api_handler(host);
    if (!handler) {
        free(host);
        return 84;
    }
    request = network_create_request_no_params_no_body((route_t) {GET, "/api/v1/test"});
    if (!request) {
        free(host);
        destroy_api_handler(handler);
        return 84;
    }
    handler_global = handler;
    network_send_request(handler, request, handle_response);
    while (handler->running) {
        network_manager_handle_waiting_sockets(handler->manager);
        usleep(100);
    }
    destroy_api_handler(handler);
    free(host);
    return 0;
}```

### CODE REVIEW

**Review by 6im0n — DISMISSED** [review]

File: tests/network/test_client.c [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    handler_global = handler;
+    network_send_request(handler, request, handle_response);
+    while (handler->running) {
+        network_manager_handle_waiting_sockets(handler->manager);
+        usleep(100);
+    }
+    destroy_api_handler(handler);
+    free(host);
+    return 0;
+}
```
Comment (Marius-P1) [inline tests/network/test_client.c]:
```suggestion
}

```

File: tests/network/test_client.c
Comment (Marius-P1) [inline tests/network/test_client.c:1]: This file is useless if it was just for test purposes

File: tests/network/test_server.c
Comment (Marius-P1) [inline tests/network/test_server.c:1]: Same for this one, you can delete it if it was just for test purposes

File: common/network/include/network/router.h [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+ * @param route The route to remove
+ */
+void network_router_remove_route(network_router_t *router, route_t route);
+/**
+ * @brief Find a route in a network router
+ * @param router The router to find the route in
+ * @param route The route to find
+ * @return The index of the route or -1 if not found
+ */
+ssize_t router_find_route(network_router_t *router, route_t route);
```
Comment (6im0n) [inline common/network/include/network/router.h:47]: Typo, or custom type ? 
image

File: common/network/include/network/router.h [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+ * @param route The route to remove the middleware from
+ */
+void network_router_remove_middleware(network_router_t *router, route_t route);
+/**
+ * @brief Find a middleware in a network router
+ * @param router The router to find the middleware in
+ * @param route The route to find the middleware for
+ * @return The index of the middleware or -1 if not found
+ */
+ssize_t router_find_middleware(network_router_t *router, route_t route);
```
Comment (6im0n) [inline common/network/include/network/router.h:83]: Same as the previous :)

**Review by Marius-P1 — DISMISSED** [review]
Juste two useless files to delete, else it's good for me :ok_hand:

File: tests/network/test_client.c
Comment (axel-eck) [reply tests/network/test_client.c:1]: I plan to add a functional testing library so this file will come handy

File: tests/network/test_server.c
Comment (axel-eck) [reply tests/network/test_server.c:1]: I plan to add a functional testing library so this file will come handy

File: common/network/include/network/router.h (hunk 3 above)
Comment (axel-eck) [reply common/network/include/network/router.h:83]: Fixed

File: common/network/include/network/router.h (hunk 2 above)
Comment (axel-eck) [reply common/network/include/network/router.h:47]: Fixed

**Review by 6im0n — DISMISSED** [review]
OK, good for me !

**Review by Marius-P1 — DISMISSED** [review]
:ok_hand:

**Review by 6im0n — APPROVED** [review]

**Review by Marius-P1 — APPROVED** [review]
:ok:

**Pull request merged**